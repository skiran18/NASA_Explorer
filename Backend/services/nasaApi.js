const axios = require('axios');
const config = require('../config');

const nasaClient = axios.create({
  baseURL: config.nasa.baseUrl,
  params: {
    api_key: config.nasa.apiKey
  }
});

class NasaApiService {
  async getAPOD() {
    const response = await nasaClient.get('/planetary/apod');
    return response.data;
  }

  async getMarsRoverPhotos(params) {
    console.log(params)
    const response = await nasaClient.get('/mars-photos/api/v1/rovers/curiosity/photos', {
      params
    });
    return response.data;
  }

  async getNearEarthObjects(params) {
    const response = await nasaClient.get('/neo/rest/v1/feed',{
      params
    });

    console.log(response)
    return response.data;
  }
}

module.exports = new NasaApiService();