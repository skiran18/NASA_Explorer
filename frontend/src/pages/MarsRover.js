import React from "react";

const MarsRover = () => {
  return (
    <div className="mars-rover">
      <h1>Mars Rover Images</h1>
      <p>Boarding pass feature and infinite scroll gallery.</p>
    </div>
  );
};

export default MarsRover;