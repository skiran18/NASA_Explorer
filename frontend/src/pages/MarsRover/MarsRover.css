/* General Container */
.mars-container {
    max-width: 90%;
    margin: 50px auto;
    text-align: center;
}

/* Camera Filter Dropdown */
.filter-section {
    margin-bottom: 20px;
    font-size: 16px;
    color: white;
}

.filter-section select {
    padding: 8px 12px;
    font-size: 14px;
    border-radius: 5px;
    border: 1px solid #ccc;
    cursor: pointer;
}

/* Grid Layout */
.photo-grid {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
    gap: 20px;
    justify-content: center;
}

/* Photo Card */
.photo-card {
    background-color: rgba(255, 255, 255, 0.1);
    border-radius: 10px;
    overflow: hidden;
    box-shadow: 0px 4px 6px rgba(0, 0, 0, 0.4);
    transition: transform 0.3s ease-in-out;
}

.photo-card:hover {
    transform: scale(1.05);
}

/* Image Styling */
.photo-card img {
    width: 100%;
    height: 200px;
    object-fit: cover;
    border-bottom: 3px solid #ffcc00;
}

/* Photo Info */
.photo-info {
    padding: 10px;
    color: white;
    font-size: 14px;
    text-align: left;
}

/* Loading Spinner */
.loading-container {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 400px;
}

.spinner {
    width: 50px;
    height: 50px;
    border: 5px solid rgba(255, 255, 255, 0.3);
    border-top: 5px solid #ffcc00;
    border-radius: 50%;
    animation: spin 1s linear infinite;
}

@keyframes spin {
    from {
        transform: rotate(0deg);
    }
    to {
        transform: rotate(360deg);
    }
}

/* No Results Message */
.no-results {
    font-size: 16px;
    color: white;
    margin-top: 20px;
}

.modal {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.6); /* Dark transparent overlay */
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 1000;
    backdrop-filter: blur(10px); /* Blurred background effect */
}

/* Modal Content */
.modal-content {
    position: relative;
    background: rgba(255, 255, 255, 0.15); /* Glass effect */
    padding: 15px;
    border-radius: 12px;
    max-width: 90%;
    max-height: 90%;
    display: flex;
    justify-content: center;
    align-items: center;
    box-shadow: 0 4px 15px rgba(0, 0, 0, 0.5);
    backdrop-filter: blur(15px);
    animation: fadeIn 0.3s ease-in-out;
}

/* Ensuring the Image Fits */
.modal-image {
    width: auto;
    height: auto;
    max-width: 100%;
    max-height: 80vh;  /* Ensures it doesn't exceed viewport height */
    object-fit: contain; /* Prevents cropping */
    border-radius: 8px;
    transition: transform 0.2s ease-in-out;
}

/* Close Button */
.close-button {
    position: absolute;
    top: 10px;
    right: 15px;
    font-size: 32px;
    font-weight: bold;
    color: white;
    cursor: pointer;
    background: none;
    border: none;
    outline: none;
    transition: color 0.3s ease;
}

.close-button:hover {
    color: red;
}

/* Animation */
@keyframes fadeIn {
    from {
        opacity: 0;
        transform: scale(0.95);
    }
    to {
        opacity: 1;
        transform: scale(1);
    }
}