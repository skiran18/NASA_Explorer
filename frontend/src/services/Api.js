const API_BASE_URL = process.env.REACT_APP_API_BASE_URL;

export const fetchData = async (endpoint) => {
  try {
    const response = await fetch(`${API_BASE_URL}/${endpoint}`);
    console.log(response)
    if (!response.ok) {
      throw new Error(`Error: ${response.statusText}`);
    }
    return await response.json();
  } catch (error) {
    console.error("API Fetch Error:", error);
    throw error;
  }
};