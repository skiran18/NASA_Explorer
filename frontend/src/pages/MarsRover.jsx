import React, { useState, useEffect } from "react";
import { fetchData } from "../services/Api";
import "./MarsRover.css";

const MarsRoverSection = () => {
  const [photos, setPhotos] = useState([]);
  const [loading, setLoading] = useState(true);
  const [selectedCamera, setSelectedCamera] = useState("ALL");
  const [modalImage, setModalImage] = useState(null);

  const cameraOptions = [
    "ALL",
    "FHAZ",
    "RHAZ",
    "MAST",
    "CHEMCAM",
    "MAHLI",
    "MARDI",
    "NAVCAM",
    "PANCAM",
    "MINITES",
  ];

  useEffect(() => {
    setLoading(true);
    let endpoint = "mars";

    if (selectedCamera !== "ALL") {
      endpoint += `/${selectedCamera}`;
    }

    fetchData(endpoint)
      .then((result) => {
        setPhotos(result.photos);
        setLoading(false);
      })
      .catch((error) => {
        console.error("Error fetching Mars photos:", error);
        setLoading(false);
      });
  }, [selectedCamera]);

  return (
    <div className={`mars-container ${modalImage ? "blurred-background" : ""}`}>
      {/* Camera Filter Dropdown */}
      <div className="filter-section">
        <label htmlFor="camera-filter">Select Camera: </label>
        <select
          id="camera-filter"
          value={selectedCamera}
          onChange={(e) => setSelectedCamera(e.target.value)}
        >
          {cameraOptions.map((camera) => (
            <option key={camera} value={camera}>
              {camera}
            </option>
          ))}
        </select>
      </div>

      {/* Loading Spinner */}
      {loading ? (
        <div className="loading-container">
          <div className="spinner"></div>
        </div>
      ) : (
        <div className="photo-grid">
          {photos.length > 0 ? (
            photos.map((photo) => (
              <div key={photo.id} className="photo-card" onClick={() => setModalImage(photo.img_src)}>
                <img src={photo.img_src} alt={`Mars Rover ${photo.id}`} />
                <div className="photo-info">
                  <p><strong>Camera:</strong> {photo.camera.full_name}</p>
                  <p><strong>Date:</strong> {photo.earth_date}</p>
                  <p><strong>Rover:</strong> {photo.rover.name}</p>
                </div>
              </div>
            ))
          ) : (
            <p className="no-results">No photos available for this camera.</p>
          )}
        </div>
      )}

      {/* Modal Dialog Box */}
      {modalImage && (
        <div className="modal" onClick={() => setModalImage(null)}>
          <div className="modal-content" onClick={(e) => e.stopPropagation()}>
            <span className="close-button" onClick={() => setModalImage(null)}>Ã—</span>
            <img src={modalImage} alt="Mars Rover" className="modal-image" />
          </div>
        </div>
      )}
    </div>
  );
};

export default MarsRoverSection;
