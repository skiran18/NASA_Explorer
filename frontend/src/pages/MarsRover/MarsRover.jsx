import React, { useState, useEffect } from "react";
import "./MarsRover.css";
import { fetchData } from "../../services/Api";

const MarsRover = () => {
  const [name, setName] = useState("");
  const [location, setLocation] = useState("");
  const [submitted, setSubmitted] = useState(false);
  const [loading, setLoading] = useState(false);
  const [photos, setPhotos] = useState([]);
  const [selectedCamera, setSelectedCamera] = useState("ALL");
  const [modalImage, setModalImage] = useState(null);

  const cameraOptions = [
    "ALL",
    "FHAZ",
    "RHAZ",
    "MAST",
    "CHEMCAM",
    "MAHLI",
    "MARDI",
    "NAVCAM",
    "PANCAM",
    "MINITES",
  ];

  useEffect(() => {
    if (!submitted) {
      setLoading(false);
      return;
    }

    let endpoint = "mars";
    if (selectedCamera !== "ALL") {
      endpoint += `/${selectedCamera}`;
    }

    fetchData(endpoint)
      .then((result) => {
        setPhotos(result.photos);
        setLoading(false);
      })
      .catch((error) => {
        console.error("Error fetching Mars photos:", error);
        setLoading(false);
      });
  }, [selectedCamera, submitted]);

  const handleSubmit = (e) => {
    e.preventDefault();
    setSubmitted(true);
    setLoading(true);
  };

  const handleCamera = (event) => {
    const newCamera = event.target.value;
    setSelectedCamera(newCamera);
  };

  return (
    <div className={`mars-travel-container ${!submitted ? "blurred" : ""}`}>
      {!submitted && (
        <div className="dialog-box">
          <form onSubmit={handleSubmit} className="boarding-pass-form">
            <h2>Get Your Mars Boarding Pass</h2>
            <input
              type="text"
              placeholder="Enter Your Name"
              value={name}
              onChange={(e) => setName(e.target.value)}
              required
            />
            <input
              type="text"
              placeholder="Enter Your Location"
              value={location}
              onChange={(e) => setLocation(e.target.value)}
              required
            />
            <button type="submit">Launch ðŸš€</button>
          </form>
        </div>
      )}

      {loading && (
        <div className="loading-animation">
          <div className="rocket"></div>
          <p>Launching to Mars...</p>
        </div>
      )}

      {!loading && submitted && (
        <div>
          <div className="boarding-pass-container">
            <div className="boarding-pass">
              <div className="boarding-pass-header">ðŸš€ MARS BOARDING PASS</div>
              <div className="boarding-pass-body">
                <div className="passenger-details">
                  <p>
                    <strong>Passenger:</strong> {name}
                  </p>
                  <p>
                    <strong>Departure:</strong> 2025
                  </p>
                  <p>
                    <strong>From:</strong> {location}
                  </p>
                  <p>
                    <strong>Destination:</strong> Mars
                  </p>
                </div>
                <div className="barcode"></div>
                <p>Explore mars using different cameras</p>
              </div>
            </div>
          </div>

          <div className="mars-header">
            <h2 className="mars-title">ðŸ“¸ Mars Gallery</h2>
            <div className="filter-section">
              <label htmlFor="camera-filter">Select Camera: </label>
              <select
                id="camera-filter"
                value={selectedCamera}
                onChange={(e) => handleCamera(e)}
              >
                {cameraOptions.map((camera) => (
                  <option key={camera} value={camera}>
                    {camera}
                  </option>
                ))}
              </select>
            </div>
          </div>

          <div className="mars-photos">
            <div className="photo-grid">
              {photos.length > 0 ? (
                photos.map((photo) => (
                  <div
                    key={photo.id}
                    className="photo-card"
                    onClick={() => setModalImage(photo.img_src)}
                  >
                    <img src={photo.img_src} alt={`Mars Rover ${photo.id}`} />
                    <div className="photo-info">
                      <p>
                        <strong>Camera:</strong> {photo.camera.full_name}
                      </p>
                      <p>
                        <strong>Date:</strong> {photo.earth_date}
                      </p>
                      <p>
                        <strong>Rover:</strong> {photo.rover.name}
                      </p>
                    </div>
                  </div>
                ))
              ) : (
                <p className="no-results">
                  No photos available for this camera.
                </p>
              )}
            </div>
          </div>
        </div>
      )}
      {modalImage && (
        <div className="modal" onClick={() => setModalImage(null)}>
          <div className="modal-content" onClick={(e) => e.stopPropagation()}>
            <span className="close-button" onClick={() => setModalImage(null)}>
              Ã—
            </span>
            <img src={modalImage} alt="Mars Rover" className="modal-image" />
          </div>
        </div>
      )}
    </div>
  );
};

export default MarsRover;
