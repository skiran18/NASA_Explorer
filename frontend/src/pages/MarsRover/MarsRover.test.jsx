import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import MarsRover from "./MarsRover";
import { fetchData } from "../../services/Api";
import "@testing-library/jest-dom";
import { TextEncoder, TextDecoder } from "util";

global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;

// Mock fetchData to return test data
jest.mock("../../services/Api", () => ({
  fetchData: jest.fn(),
}));

beforeAll(() => {
  global.ResizeObserver = class {
    observe() {}
    unobserve() {}
    disconnect() {}
  };
});

const mockData = {
  photos: [
    {
      id: "123",
      img_src: "test.jpg",
      camera: { full_name: "Front Hazard Avoidance Camera" },
      earth_date: "2025-02-15",
      rover: { name: "Curiosity" },
    },
  ],
};

describe("MarsRover Component", () => {
  beforeEach(() => {
    fetchData.mockResolvedValue(mockData); // Mock API success response
  });

  test("MarsRover Component matches snapshot", async () => {
    const { container } = render(<MarsRover />);
    await waitFor(() => expect(screen.getByText("Select Camera:")).toBeInTheDocument());
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders the component with correct heading", async () => {
    render(<MarsRover />);
    const headingElement = await screen.findByText("Select Camera:");
    expect(headingElement).toBeInTheDocument();
  });

  test("displays Mars Rover photos correctly", async () => {
    render(<MarsRover />);
    await waitFor(() => expect(screen.getByAltText("Mars Rover 123")).toBeInTheDocument());
  });

  test("renders no photos message when API returns empty array", async () => {
    fetchData.mockResolvedValue({ photos: [] }); // Mock empty response
    render(<MarsRover />);
    await waitFor(() => expect(screen.getByText("No photos available for this camera.")).toBeInTheDocument());
  });
});