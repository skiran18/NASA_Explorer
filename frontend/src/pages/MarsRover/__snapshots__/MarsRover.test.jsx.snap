// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarsRover Component MarsRover Component matches snapshot 1`] = `
<div
  class="mars-container "
>
  <div
    class="filter-section"
  >
    <label
      for="camera-filter"
    >
      Select Camera: 
    </label>
    <select
      id="camera-filter"
    >
      <option
        value="ALL"
      >
        ALL
      </option>
      <option
        value="FHAZ"
      >
        FHAZ
      </option>
      <option
        value="RHAZ"
      >
        RHAZ
      </option>
      <option
        value="MAST"
      >
        MAST
      </option>
      <option
        value="CHEMCAM"
      >
        CHEMCAM
      </option>
      <option
        value="MAHLI"
      >
        MAHLI
      </option>
      <option
        value="MARDI"
      >
        MARDI
      </option>
      <option
        value="NAVCAM"
      >
        NAVCAM
      </option>
      <option
        value="PANCAM"
      >
        PANCAM
      </option>
      <option
        value="MINITES"
      >
        MINITES
      </option>
    </select>
  </div>
  <div
    class="photo-grid"
  >
    <div
      class="photo-card"
    >
      <img
        alt="Mars Rover 123"
        src="test.jpg"
      />
      <div
        class="photo-info"
      >
        <p>
          <strong>
            Camera:
          </strong>
           
          Front Hazard Avoidance Camera
        </p>
        <p>
          <strong>
            Date:
          </strong>
           
          2025-02-15
        </p>
        <p>
          <strong>
            Rover:
          </strong>
           
          Curiosity
        </p>
      </div>
    </div>
  </div>
</div>
`;
